### PR TITLE
[flutter_tools] Instantiate shutdown hooks before localfilesystem

### LINKED ISSUE #84094: [flutter_tools] Tool generates loads of temp files on windows
I have been using Flutter for over a year, but recently I have noticed that the computer's drive is being filled with temp files caused when the flutter run is called. Every few days I clear the temp folder which is around 5gb.

Flutter doctor reports no issues:

Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 2.2.0, on Microsoft Windows [Version 10.0.19042.985], locale en-GB)
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[√] Chrome - develop for the web
[√] Android Studio (version 4.1.0)
[√] VS Code (version 1.56.2)
[√] Connected device (2 available)

Is this normal?

image

### PR DESCRIPTION
**Pull request by christopherfujino**
Fixes https://github.com/flutter/flutter/issues/84094

The core issue was that the tool was constructing the `LocalFileSystem` **BEFORE** the context runner, because it is needed to determine the `Cache.flutterRoot`, which many other objects depend on. While constructing the `LocalFileSystem`, the context getter for `ShutdownHooks` was called; this would return `null` at this point, as the context override was not set yet. This meant that later, when the `LocalFileSystem` would create the `systemTempDirectory` (which was actually a sub-directory name-spaced with `flutter_tools.`), it would try to register a shutdown hook to delete it; however, this would be a no-op since `LocalFileSystem._shutdownHooks` was always `null`.

The fix was to make `globals.shutdownHooks` a non-nullable global field and `LocalFileSystem.shutdownHooks` likewise non-nullable.

I will have a follow-up PR that will help users clean up the extra directories already created, as on Windows the OS will never garbage collect them.

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/file_system.dart [hunk 1]
```diff
@@ -206,7 +208,7 @@ class LocalFileSystem extends local_fs.LocalFileSystem {
         _systemTemp?.deleteSync(recursive: true);
       }
     } on FileSystemException {
-      // ignore.
+      globals.printTrace('Failed trying to clean up temporary directory ${_systemTemp?.path}.');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/file_system.dart:211]: Logger cannot be direct injected when this class is instantiated as it is instantiated before the logger.

File: packages/flutter_tools/lib/src/base/process.dart [hunk 2]
```diff
@@ -13,7 +13,7 @@ import 'logger.dart';
 typedef StringConverter = String? Function(String string);
 
 /// A function that will be run before the VM exits.
-typedef ShutdownHook = FutureOr<dynamic> Function();
+typedef ShutdownHook = FutureOr<void> Function();
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/process.dart:16]: the return value is never used.

File: packages/flutter_tools/lib/src/base/process.dart [hunk 3]
```diff
@@ -22,11 +22,7 @@ typedef ShutdownHook = FutureOr<dynamic> Function();
 // for more details.
 
 abstract class ShutdownHooks {
-  factory ShutdownHooks({
-    required Logger logger,
-  }) => _DefaultShutdownHooks(
-    logger: logger,
-  );
+  factory ShutdownHooks() => _DefaultShutdownHooks();
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/process.dart:25]: Now that we construct this class before the context_runner, delay passing the Logger until `runShutdownHooks` is called.

File: packages/flutter_tools/lib/src/base/process.dart [hunk 4]
```diff
@@ -40,15 +36,12 @@ abstract class ShutdownHooks {
   /// hooks within a given stage will be started in parallel and will be
   /// guaranteed to run to completion before shutdown hooks in the next stage are
   /// started.
-  Future<void> runShutdownHooks();
+  Future<void> runShutdownHooks(Logger logger);
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/base/process.dart:39]: Could use a comment for why the Logger is passed in here and not the factory.

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart [hunk 5]
```diff
@@ -384,7 +384,8 @@ void main() {
             return null;
           }),
           Barrier('Performing hot restart...'.padRight(progressMessageWidth)),
-          Multiple(<Pattern>[RegExp(r'^Restarted application in [0-9]+ms.$'), 'called main', 'called paint'], handler: (String line) {
+          // This could look like 'Restarted application in 1,237ms.'
+          Multiple(<Pattern>[RegExp(r'^Restarted application in [0-9][0-9,]*ms.$'), 'called main', 'called paint'], handler: (String line) {
```
Comment (jmagman) [inline packages/flutter_tools/test/integration.shard/overall_experience_test.dart:388]: Could also handle `10.1s` by not getting cute with the numbers.
```
^Restarted application in.*s.$
```

File: packages/flutter_tools/lib/src/base/file_system.dart [hunk 6]
```diff
@@ -239,7 +239,7 @@ class LocalFileSystem extends local_fs.LocalFileSystem {
       }
       // Make sure that the temporary directory is cleaned up when the tool
       // exits normally.
-      _shutdownHooks?.addShutdownHook(
+      _shutdownHooks.addShutdownHook(
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/file_system.dart:242]: A test that `_tryToDeleteTemp` runs at some point would be good. I wouldn't be surprised if there's already a test that's trying to do that, and it just wasn't working for some reason.

**Review by zanderso — COMMENTED** [review]
Great catch, Chris!

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart (hunk 5 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:388]: will do...however I'd be happy to have this test fail if we could hot restart that fast

File: packages/flutter_tools/lib/src/base/file_system.dart (hunk 6 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/file_system.dart:242]: You are right, there were tests for this that didn't verify it, however I think I can think of one now...

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/base/process.dart (hunk 4 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/process.dart:46]: done

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart (hunk 5 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:388]: oh never mind, this is seconds, I see what you mean

Comment (christopherfujino) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart]: done

**Comment by eggzotic** [comment]
So, on Windows 11, `Flutter 3.10.6 • channel stable`, I still get a new `flutter_tools.*` dir of ~4 GB for every launch of `flutter run ...` - both from the git bash cmd-line and when run via VS Code's run-and-debug option. Is there something still awry that causes this garbage to amass in my user's `AppData/Local/Temp/` folder ?

**Comment by zanderso** [comment]
Hi @eggzotic, we typically don't triage comments on old PRs. Would you mind filing a new GitHub issue with full details so we can properly triage it? Thanks!

**Pull request merged**